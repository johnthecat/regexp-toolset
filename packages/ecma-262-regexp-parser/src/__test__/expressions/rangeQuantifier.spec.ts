import { expect, describe, it } from 'vitest';
import { parseRegexp } from '../../index.js';

/**
 * @see parseRangeQuantifier
 */
describe('Range quantifier', () => {
  describe.each([
    { type: 'exact', quantifier: '{2}' },
    { type: 'from', quantifier: '{2,}' },
    { type: 'from to', quantifier: '{2,3}' },
  ])('$type (a$quantifier)', ({ quantifier }) => {
    it('should parse repetition count', () => {
      expect(parseRegexp(new RegExp(`a${quantifier}`))).toMatchSnapshot();
    });

    it('should correct work with lazy quantifier', () => {
      expect(parseRegexp(new RegExp(`a${quantifier}?`))).toMatchSnapshot();
    });
  });

  describe('Special cases', () => {
    it('should throw if range start is bigger than end', () => {
      expect(() => parseRegexp('/a{4,2}/')).toThrowErrorMatchingInlineSnapshot(`
        "
         ❱ /a{4,2}/
             ═════
         The quantifier range is out of order"
      `);
    });
  });

  describe('common errors', () => {
    it('should detect to quantifiable tokens', () => {
      expect(() => parseRegexp('/${1}/')).toThrowErrorMatchingInlineSnapshot(`
        "
         ❱ /\${1}/
            ↑
         The preceding token is not quantifiable"
      `);
    });

    it('should throw if there is no expression before', () => {
      expect(() => parseRegexp('/{1}/')).toThrowErrorMatchingInlineSnapshot(`
        "
         ❱ /{1}/
            ↑
         There is nothing to quantify"
      `);
    });
  });
});
