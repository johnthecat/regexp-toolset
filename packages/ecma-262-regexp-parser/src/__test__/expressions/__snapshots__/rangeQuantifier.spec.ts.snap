// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Range quantifier > 'exact' (a'{2}') > should correct work with lazy quantifier 1`] = `
{
  "body": {
    "end": 5,
    "expression": {
      "charCode": 97,
      "end": 1,
      "kind": "Char",
      "start": 1,
      "type": "Simple",
      "value": "a",
    },
    "kind": "Repetition",
    "quantifier": {
      "end": 5,
      "from": 2,
      "greedy": false,
      "kind": "Quantifier",
      "start": 2,
      "to": undefined,
      "type": 3,
    },
    "start": 1,
  },
  "end": 6,
  "flags": "",
  "kind": "Regexp",
  "start": 0,
}
`;

exports[`Range quantifier > 'exact' (a'{2}') > should parse repetition count 1`] = `
{
  "body": {
    "end": 4,
    "expression": {
      "charCode": 97,
      "end": 1,
      "kind": "Char",
      "start": 1,
      "type": "Simple",
      "value": "a",
    },
    "kind": "Repetition",
    "quantifier": {
      "end": 4,
      "from": 2,
      "greedy": true,
      "kind": "Quantifier",
      "start": 2,
      "to": undefined,
      "type": 3,
    },
    "start": 1,
  },
  "end": 5,
  "flags": "",
  "kind": "Regexp",
  "start": 0,
}
`;

exports[`Range quantifier > 'from to' (a'{2,3}') > should correct work with lazy quantifier 1`] = `
{
  "body": {
    "end": 7,
    "expression": {
      "charCode": 97,
      "end": 1,
      "kind": "Char",
      "start": 1,
      "type": "Simple",
      "value": "a",
    },
    "kind": "Repetition",
    "quantifier": {
      "end": 7,
      "from": 2,
      "greedy": false,
      "kind": "Quantifier",
      "start": 2,
      "to": 3,
      "type": 3,
    },
    "start": 1,
  },
  "end": 8,
  "flags": "",
  "kind": "Regexp",
  "start": 0,
}
`;

exports[`Range quantifier > 'from to' (a'{2,3}') > should parse repetition count 1`] = `
{
  "body": {
    "end": 6,
    "expression": {
      "charCode": 97,
      "end": 1,
      "kind": "Char",
      "start": 1,
      "type": "Simple",
      "value": "a",
    },
    "kind": "Repetition",
    "quantifier": {
      "end": 6,
      "from": 2,
      "greedy": true,
      "kind": "Quantifier",
      "start": 2,
      "to": 3,
      "type": 3,
    },
    "start": 1,
  },
  "end": 7,
  "flags": "",
  "kind": "Regexp",
  "start": 0,
}
`;

exports[`Range quantifier > 'from' (a'{2,}') > should correct work with lazy quantifier 1`] = `
{
  "body": {
    "end": 6,
    "expression": {
      "charCode": 97,
      "end": 1,
      "kind": "Char",
      "start": 1,
      "type": "Simple",
      "value": "a",
    },
    "kind": "Repetition",
    "quantifier": {
      "end": 6,
      "from": 2,
      "greedy": false,
      "kind": "Quantifier",
      "start": 2,
      "to": 9007199254740991,
      "type": 3,
    },
    "start": 1,
  },
  "end": 7,
  "flags": "",
  "kind": "Regexp",
  "start": 0,
}
`;

exports[`Range quantifier > 'from' (a'{2,}') > should parse repetition count 1`] = `
{
  "body": {
    "end": 5,
    "expression": {
      "charCode": 97,
      "end": 1,
      "kind": "Char",
      "start": 1,
      "type": "Simple",
      "value": "a",
    },
    "kind": "Repetition",
    "quantifier": {
      "end": 5,
      "from": 2,
      "greedy": true,
      "kind": "Quantifier",
      "start": 2,
      "to": 9007199254740991,
      "type": 3,
    },
    "start": 1,
  },
  "end": 6,
  "flags": "",
  "kind": "Regexp",
  "start": 0,
}
`;
